### PR TITLE
Protect consensus message buffer, prevent unlimited message

### PR DESCRIPTION
**Pull request by ShengguangXiao**
## Description

Protect consensus message buffer, prevent unlimited message

Check the sender id and ip address is in my committee member list, and check if the same message already saved in the buffer before add it.

From security audit report issue 32 and 33.
`Node::ProcessMicroBlockConsensus`
    - an attacker can cause an out-of-memory condition by sending bogus Node micro block consensus messages. These are buffered in an unbounded map `m_microBlockConsensusBuffer`
`Node::ProcessMicroBlockConsensusCore`
    - an attacker can cause an arbitrary number of threads to get temporarily stuck this method. This is limited to certain phases of the consensus protocol

### Implementation

- [x] **ready for review**

### Integration Test (Core Team)

- [x] local machine test

- [x] small-scale cloud test

### CODE REVIEW

**Comment by ansnunez** [comment]
@ShengguangXiao I think you misinterpreted the purpose of `GetConsensusID`.  It's supposed to be the ID of the ongoing consensus, not the ID of any of the nodes (leader or backup).  So changing it to `GetLeaderConsensusID` and `GetBackupConsensusID` is not in line with the purpose.

**Comment by ShengguangXiao** [comment]
> GetConsensusID
I use GetLeaderConsensusID means to get the leader's consensus ID, not the ID of leader, for backup is same. Seperate GetConsensusID into two functions is to use different checking.

**Comment by ansnunez** [comment]
> > GetConsensusID
> > I use GetLeaderConsensusID means to get the leader's consensus ID, not the ID of leader, for backup is same. Seperate GetConsensusID into two functions is to use different checking.

Ah understood.  You still get consensus ID as before, but also check the leader/backup ID.

File: src/libNode/MicroBlockPostProcessing.cpp [hunk 1]
```diff
@@ -143,6 +137,35 @@ void Node::CommitMicroBlockConsensusBuffer() {
   }
 }
 
+void Node::AddToMicroBlockConsensusBuffer(uint32_t consensusId,
+                                          const bytes& message,
+                                          unsigned int offset,
+                                          const Peer& peer) {
+  lock_guard<mutex> h(m_mutexMicroBlockConsensusBuffer);
+  auto& vecPeerMsg = m_microBlockConsensusBuffer[consensusId];
+  const auto consensusMsgType = message[offset];
```
Comment (ansnunez) [inline src/libNode/MicroBlockPostProcessing.cpp]: Same comment here, add size check

File: src/libDirectoryService/FinalBlockPostProcessing.cpp [hunk 2]
```diff
@@ -370,7 +362,36 @@ void DirectoryService::CommitFinalBlockConsensusBuffer() {
...
 }
 
-void DirectoryService::CleanFinalblockConsensusBuffer() {
+void DirectoryService::AddToFinalBlockConsensusBuffer(uint32_t consensusId,
+                                                      const bytes& message,
+                                                      unsigned int offset,
+                                                      const Peer& peer) {
+  lock_guard<mutex> h(m_mutexFinalBlockConsensusBuffer);
+  auto& vecPeerMsg = m_finalBlockConsensusBuffer[consensusId];
+  const auto consensusMsgType = message[offset];
```
[the remark this answers is not in the document]
Comment (ShengguangXiao) [reply src/libDirectoryService/FinalBlockPostProcessing.cpp]: In GetConsensID function, the size of the message checked already, but here can do a double check i think

File: src/libNode/MicroBlockPostProcessing.cpp (hunk 1 above)
Comment (ShengguangXiao) [reply src/libNode/MicroBlockPostProcessing.cpp]: Added check already

File: src/libMessage/Messenger.h [hunk 3]
```diff
@@ -650,6 +660,81 @@ class Messenger {
...
+                               << leaderId
+                               << " is out of my committee member size "
+                               << commNodes.size());
+      return false;
+    }
+
+    if (commNodes.end() ==
+        std::find_if(commNodes.begin(), commNodes.end(),
+                     [from](const PairOfNode& pairOfNode) {
+                       return from.GetIpAddress() ==
```
Comment (sandipbhoir) [inline src/libMessage/Messenger.h]: There could be different sender using same ip. Can use pubkey ?

**Pull request merged**